File: PoseDetectionService/pose_detection_class.py
import cv2
import mediapipe as mp

class PoseProcessor:
    def __init__(self):
        self.mp_pose = mp.solutions.pose # type: ignore
        self.pose = self.mp_pose.Pose()
        self.cap = cv2.VideoCapture(0)
        self.right_punch = False
        self.left_punch = False
        self.block = False
        self.left_dodge = False
        self.right_dodge = False
        self.center_x = 0.5
        self.center_y = 0.5
        self.max_valid_value = 1
        self.min_valid_value = 0
        self.data = []
        self.type = ""

    def process_frame(self, frame):
        if frame is None or frame.size == 0:
            return None

        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        results = self.pose.process(frame_rgb)
        if results.pose_landmarks:
            return results.pose_landmarks
        else:
            print("No pose landmarks found in the frame.")
            return None


    def draw_lines(self, frame, landmarks):
        lines = [
        (self.mp_pose.PoseLandmark.LEFT_SHOULDER, self.mp_pose.PoseLandmark.LEFT_ELBOW),
        (self.mp_pose.PoseLandmark.LEFT_ELBOW, self.mp_pose.PoseLandmark.LEFT_WRIST),
        (self.mp_pose.PoseLandmark.RIGHT_SHOULDER, self.mp_pose.PoseLandmark.RIGHT_ELBOW),
        (self.mp_pose.PoseLandmark.RIGHT_ELBOW, self.mp_pose.PoseLandmark.RIGHT_WRIST),
        (self.mp_pose.PoseLandmark.LEFT_SHOULDER, self.mp_pose.PoseLandmark.RIGHT_SHOULDER),
        (self.mp_pose.PoseLandmark.LEFT_HIP, self.mp_pose.PoseLandmark.RIGHT_HIP),
        (self.mp_pose.PoseLandmark.LEFT_SHOULDER, self.mp_pose.PoseLandmark.LEFT_HIP),
        (self.mp_pose.PoseLandmark.RIGHT_SHOULDER, self.mp_pose.PoseLandmark.RIGHT_HIP),
        (self.mp_pose.PoseLandmark.LEFT_HIP, self.mp_pose.PoseLandmark.LEFT_KNEE),
        (self.mp_pose.PoseLandmark.LEFT_KNEE, self.mp_pose.PoseLandmark.LEFT_ANKLE),
        (self.mp_pose.PoseLandmark.RIGHT_HIP, self.mp_pose.PoseLandmark.RIGHT_KNEE),
        (self.mp_pose.PoseLandmark.RIGHT_KNEE, self.mp_pose.PoseLandmark.RIGHT_ANKLE),
        (self.mp_pose.PoseLandmark.LEFT_EAR, self.mp_pose.PoseLandmark.LEFT_EYE_OUTER),
        (self.mp_pose.PoseLandmark.LEFT_EYE_INNER, self.mp_pose.PoseLandmark.LEFT_EYE_OUTER),
        (self.mp_pose.PoseLandmark.LEFT_EYE_INNER, self.mp_pose.PoseLandmark.NOSE),
        (self.mp_pose.PoseLandmark.NOSE, self.mp_pose.PoseLandmark.RIGHT_EYE_INNER),
        (self.mp_pose.PoseLandmark.RIGHT_EYE_INNER, self.mp_pose.PoseLandmark.RIGHT_EYE_OUTER),
        (self.mp_pose.PoseLandmark.RIGHT_EAR, self.mp_pose.PoseLandmark.RIGHT_EYE_OUTER),
        (self.mp_pose.PoseLandmark.MOUTH_LEFT, self.mp_pose.PoseLandmark.MOUTH_RIGHT),
    ]

        coordinates_list = []

        for line in lines:
            start_point = (
                int(landmarks[line[0]].x * frame.shape[1]),
                int(landmarks[line[0]].y * frame.shape[0]),
            )
            end_point = (
                int(landmarks[line[1]].x * frame.shape[1]),
                int(landmarks[line[1]].y * frame.shape[0]),
            )
            coordinates_list.append((start_point, end_point))
            cv2.line(frame, start_point, end_point, (0, 255, 0), 2)

        return coordinates_list


    def set_type(self, value):
        self.type = value

    def set_data(self, value):
        #Could add some filtering for error protection
        self.data = value


    def get_data(self):
         return self.data

    def get_type(self):
         return self.type
    

    def send_data(self, data, event_type):
        # Set an attribute of the class with the data and event_type
        self.set_data(data)
        self.set_type(event_type)
        #print(f"Saving data: {data}, Event type: {event_type}")

    def close_camera(self):
        self.cap.release()
        self.cv2.destroyAllWindows() # type: ignore


    def process_data(self):
        
        ret, frame = self.cap.read()
        landmarks = self.process_frame(frame)

        if landmarks:
            
                # Access normalized landmarks
                normalized_landmarks_list = landmarks.landmark

                # Draw lines on the frame using normalized coordinates
                coordinates_list = self.draw_lines(frame, normalized_landmarks_list)

                #Send the line data via the socket 
                self.send_data(coordinates_list, "lines")

                # Calculate distances using normalized coordinates
                right_hand = normalized_landmarks_list[self.mp_pose.PoseLandmark.RIGHT_WRIST.value]
                left_hand = normalized_landmarks_list[self.mp_pose.PoseLandmark.LEFT_WRIST.value]
                right_shoulder = normalized_landmarks_list[self.mp_pose.PoseLandmark.RIGHT_SHOULDER.value]
                left_shoulder = normalized_landmarks_list[self.mp_pose.PoseLandmark.LEFT_SHOULDER.value]

                distance_right_hand_shoulder = abs(right_hand.x - right_shoulder.x)
                distance_left_hand_shoulder = abs(left_hand.x - left_shoulder.x)
                distance_left_right_shoulder = abs(left_shoulder.x - right_shoulder.x)


                ## TRIGGERS

                if (self.right_punch): 
                        if((self.min_valid_value < right_hand.x < self.max_valid_value) and ((self.min_valid_value < right_hand.y < self.max_valid_value))):
                            if (distance_right_hand_shoulder > distance_left_right_shoulder):
                                self.right_punch = not self.right_punch
                                self.send_data("right punch", "event")
                else:
                        if (distance_right_hand_shoulder < distance_left_right_shoulder):
                            self.right_punch = not self.right_punch

        
                if (self.left_punch): 
                    if((self.min_valid_value < left_hand.x < self.max_valid_value) and ((self.min_valid_value < left_hand.y < self.max_valid_value))):   
                        if (distance_left_hand_shoulder > distance_left_right_shoulder):
                            self.left_punch = not self.left_punch
                            self.send_data("left punch","event")
                else:
                        if (distance_left_hand_shoulder < distance_left_right_shoulder):
                            self.left_punch = not self.left_punch

                
                if (self.block):
                        if ((right_hand.y < right_shoulder.y) and (left_hand.y < left_shoulder.y)):
                            self.block = not self.block
                            self.send_data("block","event")
                else:
                        if ((right_hand.y > right_shoulder.y) and (left_hand.y > left_shoulder.y)):
                            self.send_data("end block","event")
                            self.block = not self.block


                if (self.left_dodge):
                    if (right_shoulder.x > self.center_x):
                        self.left_dodge = not self.left_dodge
                        self.send_data("left dodge","event")
                else:
                        if (right_shoulder.x < self.center_x):
                            self.left_dodge = not self.left_dodge
                            self.send_data("end left dodge","event")


                if (self.right_dodge):
                    if (left_shoulder.x < self.center_x):
                        self.right_dodge = not self.right_dodge
                        self.send_data("right dodge","event")
                else:
                        if (left_shoulder.x > self.center_x):
                            self.right_dodge = not self.right_dodge
                            self.send_data("end right dodge","event")
                        

                # Display the frame with lines
                #cv2.imshow("Pose Lines", frame)


    def test(self):
         print("yoooo")


#pose = PoseProcessor()
#pose.process_data()

File: NodeServer/pose_detection_class.js

//import { solutions } from "@mediapipe/tasks-vision";

class PoseProcessor {
  constructor() {
    this.right_punch = false;
    this.left_punch = false;
    this.block = false;
    this.left_dodge = false;
    this.right_dodge = false;
    this.center_x = 0.5;
    this.center_y = 0.5;
    this.max_valid_value = 1;
    this.min_valid_value = 0;
    this.data = [];
    this.type = "";
  }

  processFrame(frame) {
    const poseLandmarkerResult = poseLandmarker.detect(frame);
    if (!poseLandmarkerResult.poseLandmarks) {
      return;
    }
    console.log(poseLandmarkerResult.poseLandmarks);
    return poseLandmarkerResult.poseLandmarks;
  }

  setType(value) {
    this.type = value;
  }

  setData(value) {
    this.data = value;
  }

  getData() {
    return this.data;
  }

  getType() {
    return this.type;
  }

  saveData(data, eventType) {
    this.setData(data);
    this.setType(eventType);
    // console.log(`Saving data: ${data}, Event type: ${eventType}`);
  }


  processData(landmarks) {

    if (landmarks) {
      // Draw lines on the frame using normalized coordinates
      //const coordinatesList = this.drawLines(frame, normalizedLandmarksList);

      //const coordinatesList = [];

      // for (const line of landmarks) {
      //   const startPoint = [
      //     landmarks[line[0]].x * frame.cols,
      //     landmarks[line[0]].y * frame.rows,
      //   ];
      //   const endPoint = [
      //     landmarks[line[1]].x * frame.cols,
      //     landmarks[line[1]].y * frame.rows,
      //   ];
      //   coordinatesList.push([startPoint, endPoint]);
      // }

      this.saveData(landmarks, "lines");

      // Calculate distances using normalized coordinates
      const rightHand = {x: landmarks[16].x, y: landmarks[16].y};
      const leftHand = {x: landmarks[15].x, y: landmarks[15].y}
      const rightShoulder = {x: landmarks[12].x, y: landmarks[12].y}
      const leftShoulder = {x: landmarks[11].x, y: landmarks[11].y}


      const distanceRightHandShoulder = Math.abs(rightHand.x - rightShoulder.x);
      const distanceLeftHandShoulder = Math.abs(leftHand.x - leftShoulder.x);
      const distanceLeftRightShoulder = Math.abs(leftShoulder.x - rightShoulder.x
      );



      //TRIGGERS

      if (this.right_punch) {
        if (
          this.min_valid_value < rightHand.x &&
          rightHand.x < this.max_valid_value &&
          this.min_valid_value < rightHand.y &&
          rightHand.y < this.max_valid_value
        ) {
          if (distanceRightHandShoulder > distanceLeftRightShoulder) {
            this.right_punch = !this.right_punch;
            this.saveData("right punch", "event");
          }
        }
      } else {
        if (distanceRightHandShoulder < distanceLeftRightShoulder) {
          this.right_punch = !this.right_punch;
        }
      }

      if (this.left_punch) {
        if (
          this.min_valid_value < leftHand.x &&
          leftHand.x < this.max_valid_value &&
          this.min_valid_value < leftHand.y &&
          leftHand.y < this.max_valid_value
        ) {
          if (distanceLeftHandShoulder > distanceLeftRightShoulder) {
            this.left_punch = !this.left_punch;
            this.saveData("left punch", "event");
          }
        }
      } else {
        if (distanceLeftHandShoulder < distanceLeftRightShoulder) {
          this.left_punch = !this.left_punch;
        }
      }

      if (this.block) {
        if (rightHand.y < rightShoulder.y && leftHand.y < leftShoulder.y) {
          this.block = !this.block;
          this.saveData("block", "event");
        }
      } else {
        if (rightHand.y > rightShoulder.y && leftHand.y > leftShoulder.y) {
          this.saveData("end block", "event");
          this.block = !this.block;
        }
      }

      if (this.left_dodge) {
        if (rightShoulder.x > this.center_x) {
          this.left_dodge = !this.left_dodge;
          this.saveData("left dodge", "event");
        }
      } else {
        if (rightShoulder.x < this.center_x) {
          this.left_dodge = !this.left_dodge;
          this.saveData("end left dodge", "event");
        }
      }

      if (this.right_dodge) {
        if (leftShoulder.x < this.center_x) {
          this.right_dodge = !this.right_dodge;
          this.saveData("right dodge", "event");
        }
      } else {
        if (leftShoulder.x > this.center_x) {
          this.right_dodge = !this.right_dodge;
          this.saveData("end right dodge", "event");
        }
      }

      // Display the frame with lines
      // cv.imshow("Pose Lines", frame);
    }
  }

  test() {
    console.log("yoooo");
  }

  static instance = new PoseProcessor();
}

// const pose = new PoseProcessor();
// pose.processData();

export default PoseProcessor;
